subsys/zigbee/osif/zb_nrf_serial.c

```diff
@@ -0,0 +1,478 @@
+static void uart_rx_notify(zb_bufid_t bufid)
+{
+	u8_t *rx_buf = uart_rx_buf;
+	size_t rx_buf_len = uart_rx_buf_offset;
+
+	ARG_UNUSED(bufid);
+
+	uart_rx_buf_len = 0;
+	uart_rx_buf_offset = 0;
+	uart_rx_buf = NULL;
+	k_sem_give(&rx_fill_buffer_sem);
+
+	rx_data_cb(rx_buf, rx_buf_len);
+}
+
+/**
+ * Async: handle completion of buffer reception.
+ *
+ * @param rx_buf data buffer
+ */
+static void store_rx_buf(u8_t *rx_buf)
+{
+	if (uart_rx_buf_offset >= uart_rx_buf_len) {
+		return;
+	}
+
+	/* The whole RX chunk/buffer received. */
+	if ((rx_buf == uart_rx_buf_chunk[0]) ||
+	    (rx_buf == uart_rx_buf_chunk[1])) {
+		/*
+		 * In async mode we have to consider both chunks, because both
+		 * of them were passed before recv() API was called:
+		 *  [0] - as the RX buffer. First bytes will go there.
+		 *  [1] - as the next RX buffer, as the result of
+		 *        uart_rx_enable() API call.
+		 */
+		memcpy(&uart_rx_buf[uart_rx_buf_offset],
+		       rx_buf,
+		       ZIGBEE_UART_RX_CHUNK_SIZE);
+		uart_rx_buf_offset += ZIGBEE_UART_RX_CHUNK_SIZE;
+	} else {
+		/*
+		 * The received buffer is not one of RX chunks.
+		 * The user buffer is now fully filed in with data.
+		 */
+		uart_rx_buf_offset = uart_rx_buf_len;
+	}
+}
+
+/**
+ * Async: Pass partial RX buffer in order to detect and handle RX timeout event.
+ *
+ * @param buf data buffer
+ * @param len data length.
+ *
+ * @return True if RX timeout detected.
+ */
+static bool handle_rx_timeout(u8_t *buf, size_t len)
+{
+	bool rx_timeout_occurred = false;
+
+	if ((buf == uart_rx_buf_chunk[0]) ||
+	    (buf == uart_rx_buf_chunk[1])) {
+		if (len != ZIGBEE_UART_RX_CHUNK_SIZE) {
+			rx_timeout_occurred = true;
+			memcpy(&uart_rx_buf[uart_rx_buf_offset], buf, len);
+		}
+	} else {
+		if ((uart_rx_buf_offset + len) != uart_rx_buf_len) {
+			rx_timeout_occurred = true;
+			/*
+			 * The driver was writing data directly to user's
+			 * buffer, there is no need to do copying.
+			 */
+		}
+	}
+
+	if (rx_timeout_occurred) {
+		uart_rx_buf_offset += len;
+		uart_rx_buf_len = uart_rx_buf_offset;
+		/*
+		 * Disable RX in order to reset receiver and release user buffer
+		 */
+		uart_rx_disable(uart_dev);
+	}
+
+	return rx_timeout_occurred;
+}
+
+static void uart_evt_handler(struct uart_event *evt, void *user_data)
+{
+	static bool restart_rx;
+	static u8_t *blocked_buf_chunk;
+
+	switch (evt->type) {
+	case UART_TX_DONE:
+		k_sem_give(&tx_done_sem);
+
+		if (tx_data_cb) {
+			zigbee_schedule_callback(
+				tx_data_cb,
+				SERIAL_SEND_SUCCESS);
+		}
+		break;
+
+	case UART_TX_ABORTED:
+		k_sem_give(&tx_done_sem);
+
+		if (tx_data_cb) {
+			zigbee_schedule_callback(
+				tx_data_cb,
+				SERIAL_SEND_TIMEOUT_EXPIRED);
+		}
+		break;
+
+	case UART_RX_RDY:
+		/* New chunk of data received. */
+		/* Sync: Pass received data through per-byte handler. */
+		/* Async: Since the higher logic expect the RX buffer to be
+		 *        fully filled, use UART_RX_BUF_RELEASED to inform about
+		 *        RX completion.
```

Comment:
This does not seem like a good idea. `UART_RX_BUF_RELEASED` is not always related to an RX completion. The completion should be rather detected by comparing `uart_rx_buf_offset` with `uart_rx_buf_len`.

Reply:
Comment aligned.